tsdb/agent/db.go

```diff
@@ -646,56 +660,67 @@ func (db *DB) Close() error {
 type appender struct {
 	*DB
 
-	pendingSeries  []record.RefSeries
-	pendingSamples []record.RefSample
+	pendingSeries    []record.RefSeries
+	pendingSamples   []record.RefSample
+	pendingExamplars []record.RefExemplar
 }
 
 func (a *appender) Append(ref uint64, l labels.Labels, t int64, v float64) (uint64, error) {
-	if ref == 0 {
-		return a.Add(l, t, v)
-	}
-	return ref, a.AddFast(ref, t, v)
-}
+	series := a.series.GetByID(ref)
+	if series == nil {
+		// Ensure no empty or duplicate labels have gotten through. This mirrors the
+		// equivalent validation code in the TSDB's headAppender.
+		l = l.WithoutEmpty()
+		if len(l) == 0 {
+			return 0, errors.Wrap(tsdb.ErrInvalidSample, "empty labelset")
+		}
 
-func (a *appender) Add(l labels.Labels, t int64, v float64) (uint64, error) {
-	hash := l.Hash()
-	series := a.series.GetByHash(hash, l)
-	if series != nil {
-		return series.ref, a.AddFast(series.ref, t, v)
-	}
+		if lbl, dup := l.HasDuplicateLabelNames(); dup {
+			return 0, errors.Wrap(tsdb.ErrInvalidSample, fmt.Sprintf(`label name "%s" is not unique`, lbl))
+		}
 
-	// Ensure no empty or duplicate labels have gotten through. This mirrors the
-	// equivalent validation code in the TSDB's headAppender.
-	l = l.WithoutEmpty()
-	if len(l) == 0 {
-		return 0, errors.Wrap(tsdb.ErrInvalidSample, "empty labelset")
-	}
+		var created bool
+		series, created = a.getOrCreate(l)
+		if created {
+			a.pendingSeries = append(a.pendingSeries, record.RefSeries{
+				Ref:    series.ref,
+				Labels: l,
+			})
 
-	if lbl, dup := l.HasDuplicateLabelNames(); dup {
-		return 0, errors.Wrap(tsdb.ErrInvalidSample, fmt.Sprintf(`label name "%s" is not unique`, lbl))
+			a.metrics.numActiveSeries.Inc()
+		}
 	}
 
-	ref := a.nextRef.Inc()
-	series = &memSeries{ref: ref, lset: l, lastTs: t}
+	series.Lock()
+	defer series.Unlock()
+
+	// Update last recorded timestamp. Used by Storage.gc to determine if a
+	// series is stale.
+	series.lastTs = t
```

Comment:
Shouldn't we update it in `Commit()`? (And maybe also use this time to determine OutOfOrder samples)

Reply:
Ah, that's interesting. Can you explain the reasoning a little more? I think if we did it in Commit, it would still need to happen before the call to `wal.Log`.

This would be a change from how we've been handling this for a while, so I'm trying to think about what problems this might cause. I'm mostly concerned about keeping both the GC logic and the scraper's ref ID cache intact. I don't think making this change would break either of those.

Reply:
There could be many failures between Append() and Commit(), in scraping or storage somehow. And scrapes can also expose timestamps. So if we are determining OOO with this lastTs, updating it here and rolling back would mean we cannot ingest those samples again. So we need to update it in Commit().

> I think if we did it in Commit, it would still need to happen before the call to wal.Log.

Why so?

Reply:
My initial thoughts were:

1. Would it be important for the in-memory series to have the latest timestamp prior to existing in the WAL
2. Would it be an issue if an in-memory series was created but logging it to the WAL failed

Thinking about it a little more, maybe it doesn't matter. I re-scanned through the GC code and everything should be fine if a series exists in memory but not the WAL, and future appends will also continue to work. 

Slightly related: if a series record fails to get written to the WAL, future appenders will re-use the same ref ID without ever writing a series record. We need a way to mark a series as written for handling the log error, and I think the timestamp-after-append would be a good flag here.

Reply:
For 1, I don't think it is important. It is possible to ingest out of order samples in that case, but since agent will only scrape, it is unlikely. But even if we ingest out of order samples, it is fine, the remote storage will deal with it.

For 2, nope, should not be an issue. I guess it will be garbage collected at next cycle eventually?

> Slightly related: if a series record fails to get written to the WAL, future appenders will re-use the same ref ID without ever writing a series record. We need a way to mark a series as written for handling the log error, and I think the latest timestamp would be a good flag here. (If it is >0, there was a series record written successfully)

Good point. I think the issue exists for non-agent Prometheus as well (we don't write series record later if it failed when series was created). We could definitely improve it here with the latest timestamp. (that could be a future PR and not this one if you desire)

Reply:
So you can move this line to Commit() _after_ the WAL logging. And I suggest adding out of order check based on lastTs in the Append() and also in Commit().

Reply:
Done, PTAL :)